### PR TITLE
Add  process test 

### PR DESCRIPTION
**Pull request by sat0ken**
This implements the process validation in [#361.](https://github.com/youki-dev/youki/issues/361)

### CODE REVIEW

**Comment by sat0ken** [comment]
@utam0k @YJDoc2 

I figure out why the [test](https://github.com/youki-dev/youki/actions/runs/11620925554/job/32363749775?pr=2968#step:8:253) failed.

I reproduced it on my PC.
If /test does not exist in the rootfs, the youki execution will fail as well as the E2E tests.

```
root@satoken:~/work/rootfs# ls
bin  boot  config.json  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@satoken:~/work/rootfs# youki run test
ERROR libcontainer::process::container_init_process: failed to chdir e=ENOENT
ERROR libcontainer::process::container_intermediate_process: failed to initialize container process: failed to unix syscall
ERROR libcontainer::process::container_main_process: failed to wait for init ready: exec process failed with error error in executing process : failed to unix syscall
ERROR libcontainer::container::builder_impl: failed to run container process exec process failed with error error in executing process : failed to unix syscall
ERROR youki: error in executing command: exec process failed with error error in executing process : failed to unix syscall
run failed : exec process failed with error error in executing process : failed to unix syscall
```

But runc is, even if /test is not in rootfs, runc run is working.

```
root@satoken:~/work/rootfs# ls
bin  boot  config.json  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@satoken:~/work/rootfs# runc run test
/test
```

If we add a test directory to the rootfs you use for testing, the tests will pass.

Sorry, I don't understand the detailed code of runc or youki, but it looks like runc is creating the cwd directory in [this line](https://github.com/opencontainers/runc/blob/main/libcontainer/rootfs_linux.go#L222-L228).

However, to resolve the error, maybe I need to check youki's code that prepares the rootfs?
What is the difference between the behavior of runc and youki?

**Comment by sat0ken** [comment]
@utam0k @YJDoc2 ping!
Please confirm.

https://github.com/youki-dev/youki/pull/2968#issuecomment-2451042801

File: tests/contest/runtimetest/src/main.rs [hunk 1]
```diff
@@ -20,7 +20,6 @@ fn get_spec() -> Spec {
     }
 }
 
-////////// ANCHOR: example_runtimetest_main
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/main.rs:23]: Why is this line removed here?

File: tests/contest/contest/src/tests/process/process_test.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        .build()
+        .context("failed to build spec")?;
+
+    Ok(spec)
+}
+
+fn process_test() -> TestResult {
+    let spec = test_result!(create_spec());
+    test_inside_container(spec, &|_| {
+        match fs::create_dir("/test") {
```
Comment (YJDoc2) [inline tests/contest/contest/src/tests/process/process_test.rs:26]: This should not be `/test`, you should take the bundle param in the closure and append `/test` and create that dir. This callback is run in host's fs before running the container, so you should be using bundle path and create this dir.

File: tests/contest/runtimetest/src/tests.rs [hunk 3]
```diff
@@ -1,9 +1,11 @@
+use std::env;
 use std::fs::{self, read_dir};
 use std::os::linux::fs::MetadataExt;
 use std::os::unix::fs::{FileTypeExt, PermissionsExt};
 use std::path::Path;
 
 use anyhow::{bail, Result};
+
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:8]: remove this extra blank line.

File: tests/contest/runtimetest/src/tests.rs [hunk 4]
```diff
@@ -546,6 +548,34 @@ pub fn test_io_priority_class(spec: &Spec, io_priority_class: IOPriorityClass) {
...
+    }
+
+    if env::var("testb").unwrap().to_string().ne("123") {
+        eprintln!(
+            "error due to spec environment value of testb want {:?}, got {:?}",
+            "123",
+            env::var("testb")
+        )
+    }
+}
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:577]: I think for this as well, we should cross check this with env from the spec, similar to how the go test is doing.

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
Some comments, also there are several warnings from the CI check, please take a look

**Comment by YJDoc2** [comment]
Hey @sat0ken , I took a look at the code and there are two things -
1. In the test (as per my comment), you are not creating the `/test` dir correctly. Hence youki does not find it and fails.
2. As for if to to follow or not runc's way of creating the cwd if not present, I think we should check and stick to the runtime spec, I'll check and get back to you. But if the failing test is fixed by correctly creating the dir, then maybe we should not copy this.

**Comment by YJDoc2** [comment]
Also @sat0ken a small request, please ping/comment in the tracking issue before opening a e2e test PR, so we can properly keep track. I have added your current PRs, but please comment for future ones. Thanks!

**Comment by sat0ken** [comment]
@YJDoc2 
Thank you for your comments and confirmation. I will fix the code.
I am newbie to contribution, so I apologize for not following the rules. Thank you for your continued support!

File: tests/contest/runtimetest/src/tests.rs (hunk 3 above)
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: Sorry, it was my mistake. I fixed it.

https://github.com/youki-dev/youki/pull/2968/commits/64a81a1c4e3c02a67a52b9801bc51783e28cd9e1

File: tests/contest/contest/src/tests/process/process_test.rs (hunk 2 above)
Comment (sat0ken) [reply tests/contest/contest/src/tests/process/process_test.rs]: @YJDoc2  
Sorry, I dont understand test freamwork well yet.
Can you please advise where I should create the directory?

File: tests/contest/runtimetest/src/tests.rs (hunk 4 above)
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed code like this.

https://github.com/youki-dev/youki/pull/2968/commits/bd0bbb1d707c9b329d1eb7b689c72662eb51b8c5

File: tests/contest/contest/src/tests/process/process_test.rs (hunk 2 above)
Comment (YJDoc2) [reply tests/contest/contest/src/tests/process/process_test.rs]: Hey @sat0ken , not sure if you're still working on this or done, but-
The approach you have added in https://github.com/youki-dev/youki/pull/2968/commits/988b9330955e0e1caf88befdd2fdcd9c0ec24111 is still wrong. You are getting the cwd from spec, which will still be inside the container root. What you need is basically -
```rust
 test_inside_container(spec, &|bundle| {
fs::create_dir(bundle.join("test"))
```
take `bundle` as the closure param, join test to it and then create that dir.

**Comment by YJDoc2** [comment]
Hey @sat0ken , There are some conflicts and I have left a comment above on the still failing tests, please take a look.

File: tests/contest/contest/src/tests/process/process_test.rs (hunk 2 above)
Comment (sat0ken) [reply tests/contest/contest/src/tests/process/process_test.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2968/commits/67fdf520ec714a8e3dab12f964d2bdff6a0e910d

**Comment by sat0ken** [comment]
@YJDoc2 Ping!

File: tests/contest/contest/src/tests/process/process_test.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    test_inside_container(spec, &|bundle| {
+        match fs::create_dir(bundle.join("test")) {
+            Result::Ok(_) => { /*This is expected*/ }
+            Err(e) => {
+                bail!(e)
+            }
+        }
+        let metadata = fs::metadata(bundle.join("test"))?;
+        let mut permissions = metadata.permissions();
+        permissions.set_mode(0o700);
```
Comment (YJDoc2) [inline tests/contest/contest/src/tests/process/process_test.rs:41]: Hey, what is the reason for this code? 
1. I don't think this will actually set the mode of the dir, as we are only updating the in-memory struct perms. We will need another syscall to actually set the mode, if you are trying to copy the behavior from go test
2. Also I don't get the intention behind this. We are not checking the perms in the runtimetest code, and if I'm correct then test is passing without this actually being set. So either we should test this in the runtime test code, or just remove these lines, as they essentially are no-op.

File: tests/contest/runtimetest/src/tests.rs [hunk 6]
```diff
@@ -550,6 +551,36 @@ pub fn test_io_priority_class(spec: &Spec, io_priority_class: IOPriorityClass) {
     }
 }
 
+pub fn validate_process(spec: &Spec) {
+    let process = spec.process().as_ref().unwrap();
+
+    if process.cwd().ne(&getcwd().unwrap()) {
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:557]: I think you should be able to directly use `!=` here. Also, as we are using the process.cwd and getcwd multiple times (in comparison and in the `eprintln` later, extract them as variables.

File: tests/contest/runtimetest/src/tests.rs [hunk 7]
```diff
@@ -550,6 +551,36 @@ pub fn test_io_priority_class(spec: &Spec, io_priority_class: IOPriorityClass) {
...
+    for env_str in process.env().as_ref().unwrap().iter() {
+        match env_str.split_once("=") {
+            Some((env_key, env_val)) => {
+                if env::var(env_key).unwrap() != env_val {
+                    eprintln!(
+                        "error due to spec environment value of {:?} want {:?}, got {:?}",
+                        env_key,
+                        env::var(env_key).unwrap(),
+                        env_val
+                    )
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:574]:
```suggestion
            Some((env_key, expected_val)) => {
                let actual_val = env::var(env_key).unwrap();
                if  actual_val != expected_val {
                    eprintln!(
                        "error due to mismatch of environment variable {:?} : want value {:?}, got {:?}",
                        env_key,
                        actual_val,
                        expected_val
                    )
```

Main change is to extract it as a variable, you can keep the error message same if you want.

File: tests/contest/runtimetest/src/tests.rs [hunk 8]
```diff
@@ -550,6 +551,36 @@ pub fn test_io_priority_class(spec: &Spec, io_priority_class: IOPriorityClass) {
...
+                }
+            }
+            None => {
+                eprintln!("spec env value is not correct")
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:578]:
```suggestion
                eprintln!("spec env value is not correct : expected key=value format, got {env_str}")
```

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
some nitpicks, but overall looks good.

File: tests/contest/contest/src/tests/process/process_test.rs (hunk 5 above)
Comment (sat0ken) [reply tests/contest/contest/src/tests/process/process_test.rs]: Sorry, I misunderstood and thought it was necessary after reading the runtimes-tools code.

https://github.com/opencontainers/runtime-tools/blob/master/validation/process/process.go#L20-L21

I delete code.
https://github.com/youki-dev/youki/pull/2968/commits/3099f7b774a60ef1eba4b7b68f07a88d5191967e

File: tests/contest/runtimetest/src/tests.rs (hunk 6 above)
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2968/commits/66a404ed709b82fe3618745f09dd3be2c0fde715

File: tests/contest/runtimetest/src/tests.rs (hunk 7 above)
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2968/commits/66a404ed709b82fe3618745f09dd3be2c0fde715

File: tests/contest/runtimetest/src/tests.rs (hunk 8 above)
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2968/commits/66a404ed709b82fe3618745f09dd3be2c0fde715

**Comment by sat0ken** [comment]
@YJDoc2 Thank you for review and comments. I've fixed the code. Could you please check the code?

File: tests/contest/runtimetest/src/tests.rs [hunk 9]
```diff
@@ -550,6 +551,38 @@ pub fn test_io_priority_class(spec: &Spec, io_priority_class: IOPriorityClass) {
...
+    }
+
+    for env_str in process.env().as_ref().unwrap().iter() {
+        match env_str.split_once("=") {
+            Some((env_key, expected_val)) => {
+                let actual_val = env::var(env_key).unwrap();
+                if actual_val != expected_val {
+                    eprintln!(
+                        "error due to spec environment value of {:?} want {:?}, got {:?}",
+                        env_key, actual_val, expected_val
```
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:573]:
```suggestion
                        "error due to spec environment value of {:?} want {:?}, got {:?}",
                        env_key, expected_val, actual_val
```

Comment (YJDoc2) [reply tests/contest/runtimetest/src/tests.rs]: Fixed it myself in [7a2b3be](https://github.com/youki-dev/youki/pull/2968/commits/7a2b3be277017727b030d38713c9477fa3a14df7)

**Review by YJDoc2 — APPROVED** [review]
lgtm :+1: Thanks
